crates/proof-of-sql/src/sql/ast/numerical_util.rs

```diff
@@ -0,0 +1,98 @@
+use crate::{
+    base::{
+        database::{Column, ColumnType},
+        math::decimal::{scale_scalar, Precision},
+        scalar::Scalar,
+    },
+    sql::parse::{ConversionError, ConversionResult},
+};
+use bumpalo::Bump;
+
+// For decimal type manipulation please refer to
+// https://learn.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver16
+
+/// Determine the output type of an add or subtract operation if it is possible
+/// to add or subtract the two input types. If the types are not compatible, return
+/// an error.
+pub(crate) fn try_add_subtract_column_types(
+    lhs: ColumnType,
+    rhs: ColumnType,
+) -> ConversionResult<ColumnType> {
+    if !lhs.is_numeric() || !rhs.is_numeric() {
+        return Err(ConversionError::DataTypeMismatch(
+            lhs.to_string(),
+            rhs.to_string(),
+        ));
+    }
+    if lhs.is_integer() && rhs.is_integer() {
+        // We can unwrap here because we know that both types are integers
+        return Ok(lhs.max_integer_type(&rhs).unwrap());
+    }
+    if lhs == ColumnType::Scalar || rhs == ColumnType::Scalar {
+        Ok(ColumnType::Scalar)
+    } else {
+        let left_precision_value = lhs.precision_value().unwrap_or(0) as i16;
+        let right_precision_value = rhs.precision_value().unwrap_or(0) as i16;
+        let left_scale = lhs.scale().unwrap_or(0);
+        let right_scale = rhs.scale().unwrap_or(0);
+        let scale = left_scale.max(right_scale);
+        let precision_value: i16 = scale as i16
+            + (left_precision_value - left_scale as i16)
+                .max(right_precision_value - right_scale as i16)
+            + 1_i16;
+        let precision = u8::try_from(precision_value)
+            .map_err(|_| ConversionError::InvalidPrecision(precision_value))
+            .and_then(|p| {
+                Precision::new(p).map_err(|_| ConversionError::InvalidPrecision(p as i16))
+            })?;
+        Ok(ColumnType::Decimal75(precision, scale))
+    }
+}
+
+/// Add or subtract two columns together.
+///
+/// If the columns are not compatible for addition/subtraction, return an error.
+pub(crate) fn try_add_subtract_columns<'a, S: Scalar>(
+    lhs: Column<'a, S>,
+    rhs: Column<'a, S>,
+    alloc: &'a Bump,
+    is_subtract: bool,
+) -> ConversionResult<Column<'a, S>> {
+    let lhs_len = lhs.len();
+    let rhs_len = rhs.len();
+    if lhs_len != rhs_len {
+        return Err(ConversionError::DifferentColumnLength(lhs_len, rhs_len));
+    }
```

Comment:
I think this can safely be an `assert!`, in which case this doesn't need to return a `Result`.

The main reason is that this method should never fail since it is within the prover.